pkg/proxy/proxy.go

```diff
@@ -1,267 +0,0 @@
-		headerCfg := h.Config.Authentication.Header
-		req.Header.Set(headerCfg.UserFieldName, u.User.GetName())
-		req.Header.Set(headerCfg.GroupsFieldName, strings.Join(u.User.GetGroups(), headerCfg.GroupSeparator))
-	}
-
-	return true
-}
-
-func newKubeRBACProxyAuthorizerAttributesGetter(authzConfig *authz.Config) *krpAuthorizerAttributesGetter {
-	return &krpAuthorizerAttributesGetter{authzConfig}
-}
-
-type krpAuthorizerAttributesGetter struct {
-	authzConfig *authz.Config
-}
-
-// GetRequestAttributes populates authorizer attributes for the requests to kube-rbac-proxy.
-func (n krpAuthorizerAttributesGetter) GetRequestAttributes(u user.Info, r *http.Request) []authorizer.Attributes {
-	apiVerb := ""
-	switch r.Method {
-	case "POST":
-		apiVerb = "create"
-	case "GET":
-		apiVerb = "get"
-	case "PUT":
-		apiVerb = "update"
-	case "PATCH":
-		apiVerb = "patch"
-	case "DELETE":
-		apiVerb = "delete"
-	}
-
-	var allAttrs []authorizer.Attributes
-
-	defer func() {
-		for attrs := range allAttrs {
-			klog.V(5).Infof("kube-rbac-proxy request attributes: attrs=%#+v", attrs)
-		}
-	}()
-
-	if n.authzConfig.ResourceAttributes == nil {
-		// Default attributes mirror the API attributes that would allow this access to kube-rbac-proxy
-		allAttrs := append(allAttrs, authorizer.AttributesRecord{
-			User:            u,
-			Verb:            apiVerb,
-			Namespace:       "",
-			APIGroup:        "",
-			APIVersion:      "",
-			Resource:        "",
-			Subresource:     "",
-			Name:            "",
-			ResourceRequest: false,
-			Path:            r.URL.Path,
-		})
-		return allAttrs
-	}
-
-	if n.authzConfig.Rewrites == nil {
-		allAttrs := append(allAttrs, authorizer.AttributesRecord{
-			User:            u,
-			Verb:            apiVerb,
-			Namespace:       n.authzConfig.ResourceAttributes.Namespace,
-			APIGroup:        n.authzConfig.ResourceAttributes.APIGroup,
-			APIVersion:      n.authzConfig.ResourceAttributes.APIVersion,
-			Resource:        n.authzConfig.ResourceAttributes.Resource,
-			Subresource:     n.authzConfig.ResourceAttributes.Subresource,
-			Name:            n.authzConfig.ResourceAttributes.Name,
-			ResourceRequest: true,
-		})
-		return allAttrs
-	}
-
-	params := []string{}
-	if n.authzConfig.Rewrites.ByQueryParameter != nil && n.authzConfig.Rewrites.ByQueryParameter.Name != "" {
-		if ps, ok := r.URL.Query()[n.authzConfig.Rewrites.ByQueryParameter.Name]; ok {
-			params = append(params, ps...)
-		}
-	}
-	if n.authzConfig.Rewrites.ByHTTPHeader != nil && n.authzConfig.Rewrites.ByHTTPHeader.Name != "" {
-		if p := r.Header.Get(n.authzConfig.Rewrites.ByHTTPHeader.Name); p != "" {
-			params = append(params, p)
-		}
-	}
-
-	if len(params) == 0 {
-		return allAttrs
-	}
-
-	for _, param := range params {
-		attrs := authorizer.AttributesRecord{
-			User:            u,
-			Verb:            apiVerb,
-			Namespace:       templateWithValue(n.authzConfig.ResourceAttributes.Namespace, param),
```

Comment:
So this means that the same param is passed to every possible resource attribute for golang template substitution?  That means that the header and query param can only logically control a single aspect without differentiating between header and query param?

Making sure I understand and that it's well described.  The behavior is acceptable.

Reply:
Actually, this seems somewhat risky.  If I specify multiple headers and get one to succeed, but the others fail, how can we be assured that the consuming application honors the same header or query param that passed?  

This must be documented as an "if any of these values, controlled by the caller, pass validation, the request is allowed" case.

Reply:
So as far as I understand it, 

1. We can specify values through query (multiple) and through the header (first).
2. Those values get individually applied to the config.
3. Where there are template markers, those get replaced.
4. We append it to the `authorizer.AttributesRecord`.

Reply:
TODO: pre-acceptance.  headers are map[string][]string.  This code needs to properly handle that.

Reply:
agreed